# Some Common Modules in Python.

os, sys, platform, time, datetime, help & dir, math,Random, webbrowser, calender.

The os and sys modules provide numerous tools to deal with filenames, paths, directories. The os module  contains two sub-modules os.sys (same as sys) and os.path that are dedicated to the system and directories.
We can also use the functions provided by these modules for file, directory, and path manipulations. These modules are wrappers for platform-specific modules, so functions like os.path.split work on UNIX, Windows, Mac OS, and any other platform supported by Python.

Os Module and Platform

In [3]:
# # Importing the operating system module...
# """
# The listdir() function returns the content of a directory. Note, however, that it mixes directories and files.
# """
# import os
# import os.path
# print(os.name)
# print("The os.getcwd() returns the CWD Current Working Directory.\n ",os.getcwd())   # Returns the Current Working Directory.
# print()
# # printing absolute path
# print("The os.path.abspath('.') returns the Absolute path of the CWD.\n ",os.path.abspath('.'))
# print()
# print("The os.listdir('.') returns the list of All the directories present in current directory/Folder \n ",os.listdir('.'))
# print()
# print(os.environ)

# print()
# # Importing the platform  module...
import platform
print(platform.version())
print(platform.system())
print(platform.release())
# print(platform.win32_ver())
print(platform.platform())

#1 SMP PREEMPT_DYNAMIC Fri Dec  8 16:22:06 PST 2023
Linux
6.1.55-06877-gc83437f2949f
Linux-6.1.55-06877-gc83437f2949f-x86_64-with-glibc2.31


According to the doc, os.name currently has one of the following: 
'posix', 'nt', 'os2', 'ce', 'java', 'riscos'. What you are after might be sys.platform.

The os-module lets us run different code dependent on which operating system the code is running on. 'nt' means that you are running windows, and 'posix' mac.

Accessing environment variables

In [4]:
import os
# os.environ.keys()
# os.environ[NAME] = VALUE
import os.path
print(os.path.isdir(os.getcwd()))
print(os.path.isfile("module.ipynb"))
print(os.path.exists(os.getcwd()))
print(os.path.getsize("module.ipynb"))

True
True
True
12145


mkdir() Function & rmdir() function.

In [5]:
import os
#os.mkdir('Dristol') # creates Dristol directory inside the current directory
os.rmdir('Dristol') # Removes Dristol directory inside the current directory

Renaming files or directories

In [5]:
import os, sys


# os.renames('common_modules.py', 'common_modules_renamed.py')

# listing directories
print("The dir is: %s" %os.listdir(os.getcwd()))

The dir is: ['module.ipynb']


Access to the time when a file was last modified. Nevertheless, the output is not friendly user. Under Unix it corresponds to the time since the Jan 1, 1970 (GMT) and under Mac OS since Jan 1, 1904 (GMT)Use the time module to make it easier to read:

In [8]:
import time
mtime = os.path.getmtime("module.ipynb") # returns time when the file was last modified
print(mtime)

"""
The output is not really meaningful since it is expressed in seconds.
You can use the time module to get a better layout of that time:
"""

print(time.ctime(mtime))

1703310898.8645117
Sat Dec 23 06:54:58 2023


### Python DateTime, TimeDelta, Strftime(Format) with Examples.

In Python, date, time and datetime classes provides a number of function to deal with dates, times and time intervals. Date and datetime are an object in Python,so when you manipulate them, you are actually manipulating objects and not string or timestamps. Whenever you manipulate dates or time, you need to import datetime function.

#### The datetime classes in Python are categorized into main 5 classes.

date – Manipulate just date ( Month, day, year)
time – Time independent of the day (Hour, minute, second, microsecond)
datetime – Combination of time and date (Month, day, year, hour, second, microsecond)
timedelta— A duration of time used for manipulating dates

We Learn The Following
How to Use Date & DateTime Class
Print Date using date.today()
Python Current Date and Time: now() today()
How to Format Date and Time Output with Strftime()
How to use Timedelta ObjectsPrinting Today, Date Components,Weekdays

In [9]:
from datetime import *
today = date.today()
print("Today's Date is :- ", today)
print("Date Components  :- ", today.day, today.month, today.year)
print("Today's weekdays  :- ", today.weekday())

Today's Date is :-  2023-12-22
Date Components  :-  22 12 2023
Today's weekdays  :-  4


In [10]:
# Aliasing or Renaming using as keyword
import datetime as dt
print(dt.datetime.now())

2023-12-22 21:33:39.866286


Importing Everything from an module using *
The from...import * Statement

In [11]:
from math import *
print(pi)
print(sin(60))
print(cosh(1))
print(factorial(5))

3.141592653589793
-0.3048106211022167
1.5430806348152437
120


How to Format Date and Time Output with Strftime()

In [13]:
from datetime import datetime

now= datetime.now() #get the current date and time

#%c - local date and time, %x-local's date, %X- local's time
print(now.strftime("%c"))

print(now.strftime("%x"))

print(now.strftime("%X"))

##### Time Formatting ####

#%I/%H - 12/24 Hour, %M - minute, %S - second, %p - local's AM/PM

print(now.strftime("%I:%M:%S %p")) # 12-Hour:Minute:Second:AM

print(now.strftime("%H:%M")) # 24-Hour:Minute

print(now.strftime("%I:%M")) # 24-Hour:Minute

Fri Dec 22 21:36:45 2023
12/22/23
21:36:45
09:36:45 PM
21:36
09:36


To import calendar

In [12]:
from calendar import *
year = 2023
print(calendar(year))

                                  2023

      January                   February                   March
Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su
                   1             1  2  3  4  5             1  2  3  4  5
 2  3  4  5  6  7  8       6  7  8  9 10 11 12       6  7  8  9 10 11 12
 9 10 11 12 13 14 15      13 14 15 16 17 18 19      13 14 15 16 17 18 19
16 17 18 19 20 21 22      20 21 22 23 24 25 26      20 21 22 23 24 25 26
23 24 25 26 27 28 29      27 28                     27 28 29 30 31
30 31

       April                      May                       June
Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su
                1  2       1  2  3  4  5  6  7                1  2  3  4
 3  4  5  6  7  8  9       8  9 10 11 12 13 14       5  6  7  8  9 10 11
10 11 12 13 14 15 16      15 16 17 18 19 20 21      12 13 14 15 16 17 18
17 18 19 20 21 22 23      22 23 24 25 26 27 28      19 20 21 22 23 24 25
24 25 26 27 28 29 30      

In [15]:
print('Happy New Year!')
print(calendar(2024))

Happy New Year!
                                  2024

      January                   February                   March
Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su
 1  2  3  4  5  6  7                1  2  3  4                   1  2  3
 8  9 10 11 12 13 14       5  6  7  8  9 10 11       4  5  6  7  8  9 10
15 16 17 18 19 20 21      12 13 14 15 16 17 18      11 12 13 14 15 16 17
22 23 24 25 26 27 28      19 20 21 22 23 24 25      18 19 20 21 22 23 24
29 30 31                  26 27 28 29               25 26 27 28 29 30 31

       April                      May                       June
Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su
 1  2  3  4  5  6  7             1  2  3  4  5                      1  2
 8  9 10 11 12 13 14       6  7  8  9 10 11 12       3  4  5  6  7  8  9
15 16 17 18 19 20 21      13 14 15 16 17 18 19      10 11 12 13 14 15 16
22 23 24 25 26 27 28      20 21 22 23 24 25 26      17 18 19 20 21 22 23
29 30     

In [20]:
import calendar

year = 2024
month = 1

# display the calendar
print(calendar.month(year, month))

    January 2024
Mo Tu We Th Fr Sa Su
 1  2  3  4  5  6  7
 8  9 10 11 12 13 14
15 16 17 18 19 20 21
22 23 24 25 26 27 28
29 30 31



### Operations on calendar :

calendar(year, w, l, c) :- This function displays the year, width of characters, no. of lines per week and column separations.

firstweekday() :- This function returns the first week day number. By default 0 (Monday).

In [21]:
# Python code to demonstrate the working of calendar() and firstweeksday()
 
# importing calendar module for calendar operations
import calendar
 
# using calender to print calendar of year
# prints calendar of 2024
print ("The calender of year 2024 is : ")
print (calendar.calendar(2024,2,1,6))
 

The calender of year 2024 is : 
                                  2024

      January                   February                   March
Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su
 1  2  3  4  5  6  7                1  2  3  4                   1  2  3
 8  9 10 11 12 13 14       5  6  7  8  9 10 11       4  5  6  7  8  9 10
15 16 17 18 19 20 21      12 13 14 15 16 17 18      11 12 13 14 15 16 17
22 23 24 25 26 27 28      19 20 21 22 23 24 25      18 19 20 21 22 23 24
29 30 31                  26 27 28 29               25 26 27 28 29 30 31

       April                      May                       June
Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su      Mo Tu We Th Fr Sa Su
 1  2  3  4  5  6  7             1  2  3  4  5                      1  2
 8  9 10 11 12 13 14       6  7  8  9 10 11 12       3  4  5  6  7  8  9
15 16 17 18 19 20 21      13 14 15 16 17 18 19      10 11 12 13 14 15 16
22 23 24 25 26 27 28      20 21 22 23 24 25 26      17 18 19 20 21 

In [22]:
#using firstweekday() to print starting day number
print (calendar.firstweekday())

0


- isleap (year) :- This function checks if year mentioned in argument is leap or not.
- leapdays (year1, year2) :- This function returns the number of leap days between the specified years in arguments.

In [23]:
# importing calendar module for calendar operations
import calendar
 
# using isleap() to check if year is leap or not
if (calendar.isleap(2008)):
       print ("The year is leap")
else : print ("The year is not leap")
 
#using leapdays() to print leap days between years 
print ("The leap days between 1950 and 2000 are : ",end="")
print (calendar.leapdays(1950, 2000))

The year is leap
The leap days between 1950 and 2000 are : 12


In [25]:
from datetime import *
print (timedelta(days=365, hours=8, minutes=15))

# print today's date
print ("today is: " + str(datetime.now()))

# print today's date one year from now
print ("one year from now it will be:" + str(datetime.now() + timedelta(days=365)))

365 days, 8:15:00
today is: 2023-12-23 07:23:29.412287
one year from now it will be:2024-12-22 07:23:29.412662


In [1]:
# formatted time

import time, datetime

localtime = time.asctime(time.localtime())
print ("Local current time :", localtime)

Local current time : Sat Dec 23 10:19:37 2023


# Using help() and dir()

In [5]:
dir(datetime)

['MAXYEAR',
 'MINYEAR',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'date',
 'datetime',
 'datetime_CAPI',
 'sys',
 'time',
 'timedelta',
 'timezone',
 'tzinfo']

In [7]:
help(datetime)

Help on module datetime:

NAME
    datetime - Fast implementation of the datetime type.

MODULE REFERENCE
    https://docs.python.org/3.9/library/datetime
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

CLASSES
    builtins.object
        date
            datetime
        time
        timedelta
        tzinfo
            timezone
    
    class date(builtins.object)
     |  date(year, month, day) --> date object
     |  
     |  Methods defined here:
     |  
     |  __add__(self, value, /)
     |      Return self+value.
     |  
     |  __eq__(self, value, /)
     |      Return self==value.
     |  
     |  __format__(...)
     |      Formats self with strftime.
     |  
     |  __ge__(self, value, /)
     |    

In [9]:
dir(date)

['MAXYEAR',
 'MINYEAR',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'date',
 'datetime',
 'datetime_CAPI',
 'sys',
 'time',
 'timedelta',
 'timezone',
 'tzinfo']

In [10]:
# Time module
import time
print(time.time())
import time

print(time.time())

# Unix timestamp - number of seconds from the begining of time(epoch)
# 1st Jan, 1970 @ 00:00:00

1703323652.8045585


In [11]:
# see time, sec to original format

print(time.ctime(1527664773.8468134))

# ctime(sec)

Wed May 30 08:19:33 2018


In [12]:
#ask time

print(time.asctime())

Sat Dec 23 10:28:47 2023


In [13]:
print(time.gmtime())

time.struct_time(tm_year=2023, tm_mon=12, tm_mday=23, tm_hour=9, tm_min=29, tm_sec=14, tm_wday=5, tm_yday=357, tm_isdst=0)


### Maths module

It provides access to the mathematical functions defined by the C standard.

ceil(x)	Returns the smallest integer greater than or equal to x.
copysign(x, y)	Returns x with the sign of y
fabs(x)	Returns the absolute value of x
factorial(x)	Returns the factorial of x
floor(x)	Returns the largest integer less than or equal to x
fmod(x, y)	Returns the remainder when x is divided by y
frexp(x)	Returns the mantissa and exponent of x as the pair (m, e)
fsum(iterable)	Returns an accurate floating point sum of values in the iterable
isfinite(x)	Returns True if x is neither an infinity nor a NaN (Not a Number)
isinf(x)	Returns True if x is a positive or negative infinity
isnan(x)	Returns True if x is a NaN
ldexp(x, i)	Returns x * (2**i)
modf(x)	Returns the fractional and integer parts of x
trunc(x)	Returns the truncated integer value of x
exp(x)	Returns e**x
expm1(x)	Returns e**x - 1
log(x[, base])	Returns the logarithm of x to the base (defaults to e)
log1p(x)	Returns the natural logarithm of 1+x
log2(x)	Returns the base-2 logarithm of x
log10(x)	Returns the base-10 logarithm of x
pow(x, y)	Returns x raised to the power y
sqrt(x)	Returns the square root of x
acos(x)	Returns the arc cosine of x
asin(x)	Returns the arc sine of x
atan(x)	Returns the arc tangent of x
atan2(y, x)	Returns atan(y / x)
cos(x)	Returns the cosine of x
hypot(x, y)	Returns the Euclidean norm, sqrt(x*x + y*y)
sin(x)	Returns the sine of x
tan(x)	Returns the tangent of x
degrees(x)	Converts angle x from radians to degrees
radians(x)	Converts angle x from degrees to radians
acosh(x)	Returns the inverse hyperbolic cosine of x
asinh(x)	Returns the inverse hyperbolic sine of x
atanh(x)	Returns the inverse hyperbolic tangent of x
cosh(x)	Returns the hyperbolic cosine of x
sinh(x)	Returns the hyperbolic cosine of x
tanh(x)	Returns the hyperbolic tangent of x
erf(x)	Returns the error function at x
erfc(x)	Returns the complementary error function at x
gamma(x)	Returns the Gamma function at x
lgamma(x)	Returns the natural logarithm of the absolute value of the Gamma function at x
pi	Mathematical constant, the ratio of circumference of a circle to it's diameter (3.14159...)
e	mathematical constant e (2.71828...)


In [15]:
import math
print(math.ceil(4))
print(math.ceil(4.5))
print(math.fabs(-3))
print(math.factorial(3))
print(math.floor(3.5))
print(math.fsum([0.14,0.1,0.1])) #Floating sum with precision
print(math.gcd(4,8))
print(math.exp(3))
print(math.log(8,2))
print(math.pow(2,3))


4
5
3.0
6
3
0.34
4
20.085536923187668
3.0
8.0


In [16]:
print(math.cos(0))
print(math.sin(0))
print(math.tan(0))
print(math.acos(0))
print(math.asin(0))
print(math.atan(0))
print(math.pi)
print(math.e)
print(math.tau)
print(math.inf)
print(math.nan)
f=math.nan
print(float("inf")-float("inf"))

1.0
0.0
0.0
1.5707963267948966
0.0
0.0
3.141592653589793
2.718281828459045
6.283185307179586
inf
nan
nan


Random module
When to use it ..?? --> We want the computer to pick a random number in a given range

In [17]:
from random import *

print(random())

# 0 and 1 (not inclusive)

0.04382312676967659


Random functions
The Random module contains some very useful functions
randint(a,b)
If we wanted a random integer, we can use the randint function

Randint accepts two parameters: a lowest and a highest number.

a and b inclusive

In [19]:
import random
print(random.randint(0, 5))

# 0 and 5 are inclusive (Included)

2


uniform(a, b)
Return a random floating point number between a and b inclusive

In [21]:
print(uniform(1,19))

2.3604010272484715


Choice
it Generates a random value from the sequence sequence.

The choice function can often be used for choosing a random element from a list.

In [23]:
import random
myList = [2, 109, False, 10, "Mukesh", 482, "Acadview"]
print(random.choice(myList))

Acadview


Shuffle
The shuffle function, shuffles the elements in list in place, so they are in a
random order.

In [24]:
import random

s=list(range(5))
print("List Before Shuffling is ", s)
random.shuffle(s) # << shuffle before print or assignment
print("List after Shuffling is ",s)

List Before Shuffling is  [0, 1, 2, 3, 4]
List after Shuffling is  [2, 4, 0, 3, 1]


Randrange
Generate a randomly selected element from range(start, stop, step)

Syntax :- random.randrange(start, stop[, step])

In [25]:
import random
for i in range(5):
    print(random.randrange(0, 101, 5))

35
75
45
45
100


sample(population, k)

Return a k length list of unique elements chosen from the population sequence

In [26]:
import random
c = list(range(0, 15))
print(c)

print(random.sample(c, 5))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
[8, 10, 0, 7, 3]


## Webbrowser module

The webbrowser module in Python provides an interface to display Web-based documents.

calling the open() function from this module will do the right thing.

Using webbrowser.open_new_tab A new browser tab will open with the search term in googles search field.

In [27]:
import webbrowser
google = input('Enter Your Google search:')
webbrowser.open_new_tab('http://www.google.com/search?btnG=1&q=%s' % google)

True

LITERALS ----->
    %a - abbreviated weekday name

    %A - full weekday name

    %b - abbreviated month name

    %B - full month name

    %c - preferred date and time representation

    %C - century number (the year divided by 100, range 00 to 99)

    %d - day of the month (01 to 31)

    %D - same as %m/%d/%y

    %e - day of the month (1 to 31)

    %g - like %G, but without the century

    %G - 4-digit year corresponding to the ISO week number (see %V).

    %h - same as %b

    %H - hour, using a 24-hour clock (00 to 23)

    %I - hour, using a 12-hour clock (01 to 12)

    %j - day of the year (001 to 366)

    %m - month (01 to 12)

    %M - minute

    %n - newline character

    %p - either am or pm according to the given time value

    %r - time in a.m. and p.m. notation

    %R - time in 24 hour notation

    %S - second

    %t - tab character

    %T - current time, equal to %H:%M:%S

    %u - weekday as a number (1 to 7), Monday=1. Warning: In Sun Solaris Sunday=1

    %U - week number of the current year, starting with the first Sunday as the first day of the first week

    %V - The ISO 8601 week number of the current year (01 to 53), where week 1 is the first week that has at least 4 days in the current year, and with Monday as the first day of the week

    %W - week number of the current year, starting with the first Monday as the first day of the first week

    %w - day of the week as a decimal, Sunday=0

    %x - preferred date representation without the time

    %X - preferred time representation without the date

    %y - year without a century (range 00 to 99)

    %Y - year including the century

    %Z or %z - time zone or name or abbreviation

    %% - a literal % character
